### PR TITLE
topgrade: add module

### PR DESCRIPTION
**Pull request by msfjarvis**
### Description

[topgrade](https://github.com/r-darwish/topgrade) is a CLI tool written in Rust that can detect and upgrade packages across nearly all package managers that can be present on a device.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

**Comment by teto** [comment]
looks like nix nemesis xD

**Comment by msfjarvis** [comment]
Hey @rycee, can you please review this? Thanks!

File: modules/programs/topgrade.nix [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+let
+
+  cfg = config.programs.topgrade;
+
+  tomlFormat = pkgs.formats.toml { };
+
+in {
+
+  meta.maintainers = [ maintainers.msfjarvis ];
```
Comment (berbiche) [inline modules/programs/topgrade.nix:13]:
```suggestion
  meta.maintainers = [ hm.maintainers.msfjarvis ];
```

`lib.maintainers` is for maintainers in NixOS' nixpkgs and `lib.hm.maintainers` for HM's maintainers file.

You can add yourself to nixpkgs' maintainers list without maintaining a package IIRC. You only need to open a PR 😃

File: modules/programs/topgrade.nix [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    settings = mkOption {
+      type = with types;
+        let
+          prim = either bool (either int str);
+          primOrPrimAttrs = either prim (attrsOf prim);
+          entry = either prim (listOf primOrPrimAttrs);
+          entryOrAttrsOf = t: either entry (attrsOf t);
+          entries = entryOrAttrsOf (entryOrAttrsOf entry);
+        in attrsOf entries // { description = "topgrade configuration"; };
+      default = { };
```
Comment (berbiche) [inline modules/programs/topgrade.nix:34]:
```suggestion
      default = { };
      defaultText = literalExample "{ }";
```

**Review by berbiche — CHANGES_REQUESTED** [review]
Thanks for your first contribution!

The module is very clean :+1:

File: modules/programs/topgrade.nix (hunk 1 above)
Comment (msfjarvis) [reply modules/programs/topgrade.nix]: All fixed, thanks for your review!

**Comment by berbiche** [comment]
Can you add a news entry to `modules/misc/news.nix`?

**Comment by msfjarvis** [comment]
> Can you add a news entry to `modules/misc/news.nix`?

Done.

File: modules/programs/topgrade.nix [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    settings = mkOption {
+      type = with types;
+        let
+          prim = oneOf [ bool int str ];
+          primOrPrimAttrs = either prim (attrsOf prim);
+          entry = either prim (listOf primOrPrimAttrs);
+          entryOrAttrsOf = t: either entry (attrsOf t);
+          entries = entryOrAttrsOf (entryOrAttrsOf entry);
+        in attrsOf entries // { description = "topgrade configuration"; };
```
Comment (berbiche) [inline modules/programs/topgrade.nix:33]: I just realized `tomlFormat.generate` was used.
Is there a difference between this manual type and `tomlFormat.type`?

Comment (msfjarvis) [reply modules/programs/topgrade.nix:33]: I'm not 100% sure, I based the module off the existing [starship](https://github.com/nix-community/home-manager/blob/42bb553544420942415257fb2d3938698b27d72d/modules/programs/starship.nix) configuration. Open to going with whatever's the recommended option.

Comment (berbiche) [reply modules/programs/topgrade.nix:33]: I recommend using `tomlFormat.type` in that case :+1:

See this test case that shows the resulting format:
https://github.com/NixOS/nixpkgs/blob/d600f006643e074c2ef1d72e462e218b647a096c/pkgs/pkgs-lib/tests/formats.nix#L127-L154

Comment (msfjarvis) [reply modules/programs/topgrade.nix]: Done, tests passed without any changes required 🎉

**Review by berbiche — APPROVED** [review]

**Pull request merged**